examples/rpc/main.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"golang.ngrok.com/ngrok/v2"
+	"golang.ngrok.com/ngrok/v2/rpc"
+)
+
+func main() {
+	// Create an ngrok agent with RPC handler
+	agent, err := ngrok.NewAgent(
+		ngrok.WithAuthtoken(os.Getenv("NGROK_AUTHTOKEN")),
+		ngrok.WithRPCHandler(func(ctx context.Context, session ngrok.AgentSession, req rpc.Request) ([]byte, error) {
+			// Handle different RPC methods
+			switch req.Method() {
+			case rpc.StopAgentMethod:
+				log.Println("Received stop command, shutting down...")
+				// In a real application, you might want to do some cleanup
+				// Return nil error to acknowledge the command
+				return nil, nil
+
+			case rpc.RestartAgentMethod:
+				log.Println("Received restart command")
+				// Typically you'd want to implement restart logic here
+				return nil, nil
+
+			case rpc.UpdateAgentMethod:
+				log.Println("Received update command")
+				// Implement your update logic here
+				return nil, nil
+
+			default:
+				err := fmt.Errorf("unsupported method: %s", req.Method())
+				log.Println(err)
+				return nil, err
+			}
+		}),
+	)
+	if err != nil {
+		log.Fatalf("Error creating agent: %v", err)
+	}
+
+	// Connect to ngrok
+	_, err = agent.Connect(context.Background())
+	if err != nil {
+		log.Fatalf("Error connecting: %v", err)
+	}
+
+	log.Printf("Agent connected and ready to handle RPC commands")
+
+	// Create a simple TCP endpoint for demonstration
+	listener, err := agent.Listen(context.Background())
+	if err != nil {
+		log.Fatalf("Error creating endpoint: %v", err)
+	}
+	log.Printf("Endpoint created: %s", listener.URL())
+
+	// Wait for signal to exit
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+	<-sigChan
```

Comment:
We're still stuck waiting here, even if we get a `Step` request in the dashboard.

We do get an INFO level log line (`level=INFO msg="accept failed" obj=csess id=958a03312932 err="reconnecting session closed"`) for the disconnect caused by clicking 'Stop' in the dashboard, but as best I can tell there's no way to observe it on the session object or agent object.

I wanted to do something like `<-agent.Wait(ctx)` or `<-session.Wait(ctx)` to fix it...

But since we don't have that, I think we'll have to make our own shutdown channel, and plumb closing that through the `WithRPCHandler`.

Reply:
i've improved this example to show creating your own channel. see the comment below about an `agent.Wait()` or `session.Wait()` API